### PR TITLE
Protocol Specific Network Metrics - Iteration 2

### PR DESCRIPTION
**Pull request by puzpuzpuz**
* Bytes send/received metrics for each `ProtocolType` are added as metrics and logged to diagnostics (only total, not per-thread or per-pipeline).
* Bytes send/received are reported to Management Center.
* Stats are exposed in `NetworkingService` as it seems to be the most logic place for them at the moment. If more stats arise in future, a separate class might be introduced to handle all of them.

Note that this works only when advanced networking is enabled and an endpoint configuration is made for the specific protocol type (member, client, WAN etc.).

MC PR: hazelcast/management-center#2062

Previous iteration: #15486

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java [hunk 1]
```diff
@@ -102,7 +103,7 @@
     private volatile NioThread[] inputThreads;
     private volatile NioThread[] outputThreads;
 
-    // Currently this is a course grained aggregation of the bytes/send reveived.
+    // Currently this is a course grained aggregation of the bytes/send received.
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java]: While we're at it: `course` -> `coarse`

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java]: Thanks for finding this typo! I'll fix this one.

**Comment by puzpuzpuz** [comment]
@pveentjer could you review this PR on this week?

File: hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import static com.hazelcast.util.JsonUtil.getObject;
+
+/**
+ * Stats per {@link ProtocolType} for a single direction of network traffic (inbound or outbound).
+ * <p>
+ * Stores number of bytes sent or received per {@link ProtocolType} depending on whether it is
+ * for inbound or outbound traffic. Works only when Advanced Networking is enabled and
+ * {@link com.hazelcast.config.EndpointConfig} is added for the {@link ProtocolType} in question.
+ */
+@PrivateApi
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java]: You don't need to mark it private. It is in com.hz.internal. So it is private independent of this annotation

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java [hunk 3]
```diff
@@ -175,6 +186,14 @@ public synchronized void start() {
 
         networking.start();
         startAcceptor();
+
+        if (advancedNetworkingEnabled) {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java]: I need to think about this part.

File: hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java (hunk 2 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java]: I didn't know that. Going to remove the annotation.

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java (hunk 3 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java]: The idea is that users who don't have advanced network stats enabled shouldn't "pay" for them. It doesn't make sense to spend CPU cycles on collection of these stats, as we know that they're going to be zeros.

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java]: Should this not always be enabled by default?

Also: the dynamic probe discovery stuff is pending.

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java]: > Should this not always be enabled by default?

As far as I know, you need to enable advanced network config explicitly: https://docs.hazelcast.org/docs/3.12.2/manual/html-single/index.html#advanced-network-configuration

It makes sense that it's not enabled by default as advanced network config is about using separate ports for client/member/WAN/REST/memcache protocols.

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpNetworkingService.java]: > Also: the dynamic probe discovery stuff is pending.

I'm not sure how it relates with this PR. Maybe you mean metrics redesign task? If that's the case, then you should check #15560

File: hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+    /**
+     * For serializing the stats before sending to Management Center.
+     *
+     * @return the JSON representation of this object
+     */
+    @Override
+    public JsonObject toJson() {
```
Comment (tkountis) [inline hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java]: JSON un/marshalling inside this HZ layer, is a bit out of place. How about we move this logic on a dto helper inside the MC relevant package?

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/networking/nio/AdvancedNetworkStats.java]: Thanks for the heads-up! I'll move this logic into MC package.

**Comment by puzpuzpuz** [comment]
I have pushed changes that address all comments + one issue that wasn't solved in the initial implementation. Namely, per protocol stats weren't including counters for closed connections/channels, thus their counters weren't monotonic. This issue should be fixed now and covered with a new test.

**Comment by puzpuzpuz** [comment]
Leaving a follow-up note with potential enhancements in future as the result of chat with @tkountis.

We may convert `#calculate*` methods in EMs into something like `getNetStats()` (with a couple of counters for bytes received/sent). Each of these NetStats objects will be updated individually with a metrics task. Then AggregateEM will return a `Map<EndpointQualifier, NetStats>`.

This API would be more flexible and feel more natural.

**Pull request merged**